include/Sema/Constraints.hpp

```diff
@@ -0,0 +1,624 @@
+
+    ///
+    /// @brief Constructs a constraint with a conversion restriction.
+    ///
+    /// @param kind Constraint kind.
+    /// @param restriction Conversion restriction kind.
+    /// @param first Source type.
+    /// @param second Target type.
+    /// @param locator AST source node.
+    /// @param typeVars Type variables involved.
+    ///
+    Constraint(
+        ConstraintKind kind, ConversionRestrictionKind restriction,
+        glu::types::Ty first, glu::types::Ty second, glu::ast::ASTNode *locator
+    );
+
+    /// @brief Constructs a constraint with a Member.
+    /// @param kind Constraint kind.
+    /// @param first First type.
+    /// @param second Second type.
+    /// @param member The member expression.
+    /// @param locator AST source node.
+    /// @param typeVars Type variables involved.
+    Constraint(
+        ConstraintKind kind, glu::types::Ty first, glu::types::Ty second,
+        glu::ast::StructMemberExpr *member, glu::ast::ASTNode *locator
+    );
+
+    /// @brief Constructs a syntactic element constraint.
+    ///
+    /// @param node The AST node representing the syntactic element.
+    /// @param isDiscarded Whether the result of this constraint is unused.
+    /// @param locator The AST node that triggered this constraint.
+    /// @param typeVars Type variables involved in the constraint.
+    ///
+    Constraint(
+        glu::ast::ASTNode node, bool isDiscarded, glu::ast::ASTNode *locator
+    );
+
+    /// @brief Constructs a bind overload constraint.
+    ///
+    /// @param type The type to be constrained.
+    /// @param choice The function declaration representing the overload choice.
+    /// @param locator The AST node that triggered this constraint.
+    /// @param typeVars Type variables involved in the constraint.
+    ///
+    Constraint(
+        glu::types::Ty type, glu::ast::FunctionDecl *choice,
+        glu::ast::ASTNode *locator
+    );
+
+public:
+    /// @brief Create a new constraint between two types.
+    /// @param allocator The allocator for memory allocation.
+    /// @param kind The kind of constraint to create.
+    /// @param first The first type in the constraint.
+    /// @param second The second type in the constraint.
+    /// @param locator The AST node that triggered this constraint.
+    /// @param extraTypeVars Additional type variables to include.
+    /// @return A newly created constraint.
+    static Constraint *create(
+        llvm::BumpPtrAllocator &allocator, ConstraintKind kind,
+        glu::types::Ty first, glu::types::Ty second, glu::ast::ASTNode *locator
+    );
+
+    /// @brief Create a bind constraint between two types.
+    /// @param allocator The allocator for memory allocation.
+    /// @param first The first type in the bind.
+    /// @param second The second type in the bind.
+    /// @param locator The AST node that triggered this constraint.
+    /// @return A newly created bind constraint.
+    inline static Constraint *createBind(
```

Comment:
don't explicitly write the `inline` keyword... I can see that it's in the class, I don't need extra noise
> A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.

[inline specifier](https://en.cppreference.com/w/cpp/language/inline)

Reply:
your other inline function don't have the keyword either